File: Trees_Graphs/python/GraphQueue.py
from exceptions import GraphQueueEmptyException


class GraphQueue(object):
    def __init__(self):
        self.nodes = []
        self.size = 0

    def __str__(self):
        return ', '.join(nodes)

    def enqueue(self, node):
        self.nodes.append(node)
        self.size += 1

    def dequeue(self):
        if self.is_empty():
            raise GraphQueueEmptyException()

        first = self.nodes[0]
        self.nodes = self.nodes[1:]
        self.size -= 1

        return first

    def is_empty(self):
        return self.size == 0

File: Trees_Graphs/go/GraphQueue.go
package types


import (
  "fmt"
)


type GraphQueue struct {
  nodes []*GraphNode
  size int
}


func (q *GraphQueue) Enqueue(node *GraphNode) {
  q.nodes = append(q.nodes, node)
  q.size++
}


func (q *GraphQueue) IsEmpty() bool {
  return q.size == 0
}


func (q *GraphQueue) Dequeue() (*GraphNode, bool) {
  if q.IsEmpty() {
    return &GraphNode{}, false
  }

  var first GraphNode
  first = *q.nodes[0]
  q.nodes = q.nodes[1:]
  q.size--

  if q.IsEmpty() {
    return &first, false
  }

  return &first, true
}


func (q *GraphQueue) PrintGraphQueue() {
  for i := 0; i < q.size; i++ {
    fmt.Printf(q.nodes[i].Value)
    fmt.Printf(", ")
  }
  fmt.Printf("\n")
}
